src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -197,7 +197,9 @@ private void AllocateWriteHeadSynchronized(int sizeHint)
                             _writingHeadBytesBuffered = 0;
                         }
 
-                        BufferSegment newSegment = AllocateSegment(sizeHint);
+                        // Double the minimum segment size on subsequent segements
+                        int newSegmentSize = Math.Min(PipeOptions.MaximumSegmentSize, _writingHead.Capacity * 2);
```

Comment:
What pool are you testing this with?

Reply:
ArrayPool. I need to change the implementation of the PinnedBlockMemoryPool to make this work the way it's intended.